python_modules/dagster/dagster/_core/execution/asset_backfill.py

```diff
@@ -709,47 +730,71 @@ def _submit_runs_and_update_backfill_in_chunks(
-        debug_crash_flags={},
-        logger=logger,
-        run_tags={**run_tags, BACKFILL_ID_TAG: backfill_id},
-    ):
-        if submit_run_request_chunk_result is None:
-            # chunk is still processing, allow the daemon to heartbeat
-            yield None
-            continue
+    reserved_run_ids = asset_backfill_iteration_result.reserved_run_ids
+
+    run_request_execution_data_cache = {}
 
-        # Chunk has finished submitting
+    chunk_size = get_asset_backfill_run_chunk_size()
+
+    for run_request_idx, run_request in enumerate(run_requests):
+        run_id = reserved_run_ids[run_request_idx] if reserved_run_ids else None
+        try:
+            submit_asset_run(
+                run_id,
+                run_request._replace(
+                    tags={
+                        **run_request.tags,
+                        **run_tags,
+                        BACKFILL_ID_TAG: backfill_id,
+                    }
+                ),
+                run_request_idx,
+                instance,
+                workspace_process_context,
+                asset_graph,
+                run_request_execution_data_cache,
+                {},
+                logger,
+            )
+        except Exception:
+            # Write the runs that we submitted before hitting an error
+            _write_updated_backfill_data(
+                instance,
+                backfill_id,
+                updated_backfill_data,
+                asset_graph,
+                run_requests[num_submitted:],
+                asset_backfill_iteration_result.reserved_run_ids[num_submitted:],
+            )
+            raise
+        yield None
+
+        num_submitted += 1
 
-        num_submitted += len(submit_run_request_chunk_result.chunk_submitted_runs)
         updated_backfill_data: AssetBackfillData = (
             updated_backfill_data.with_run_requests_submitted(
-                [rr for (rr, _) in submit_run_request_chunk_result.chunk_submitted_runs],
+                [run_request],
                 asset_graph,
                 instance_queryer,
             )
         )
 
-        # Refetch, in case the backfill was requested for cancellation in the meantime
-        backfill = cast(PartitionBackfill, instance.get_backfill(backfill_id))
-        updated_backfill = backfill.with_asset_backfill_data(
-            updated_backfill_data,
-            dynamic_partitions_store=instance,
-            asset_graph=asset_graph,
-        ).with_submitting_run_requests(
-            run_requests[num_submitted:],
-            asset_backfill_iteration_result.reserved_run_ids[num_submitted:],
-        )
-        instance.update_backfill(updated_backfill)
+        # After each chunk or on the final request, write the updated backfill data
+        # and check to make sure we weren't interrupted
+        if (num_submitted % chunk_size == 0) or num_submitted == len(run_requests):
+            run_request_execution_data_cache = {}
```

Comment:
Previously this cache would be re-initialized every 25 runs just by virtue of being declared within submit_asset_runs_in_chunks - I kept that here for now but i'm not sure if that is needed? Was it useful to periodically reset the cache for long-running backfills maybe?

Reply:
oh you are right, I misread - i thought it was getting re-initialized on every chunk